tools/CreateFilterMappings.ps1

```diff
@@ -0,0 +1,457 @@
+{
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [string]$TestCsprojPath
+    )
+
+    $Content = Get-Content -Path $TestCsprojPath
+    $MatchedAssembly = $Content | where { $_ -match "<AssemblyName>[a-zA-Z.]*<\/AssemblyName>" }
+    if ($MatchedAssembly -ne $null)
+    {
+        return $MatchedAssembly.Trim().Trim("<AssemblyName>").Trim("</")
+    }
+}
+
+<#
+Turns a file path into a normalized key (strips out everything before "src").
+#>
+function Create-Key
+{
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [string]$FilePath
+    )
+
+    $Key = ""
+    $TempFilePath = $FilePath
+    while ($true)
+    {
+        $TempItem = Get-Item -Path $TempFilePath
+        $Name = $TempItem.Name
+        $Key = $Name + "/" + $Key
+        if ($Name -eq "src")
+        {
+            break
+        }
+
+        if ($TempItem.Parent -ne $null)
+        {
+            $TempFilePath = $TempItem.Parent.FullName
+        }
+        else
+        {
+            $TempFilePath = $TempItem.Directory.FullName
+        }
+    }
+
+    return $Key
+}
+
+<#
+Creates a mapping from a solution file to the projects it references. For example:
+
+{
+    "C:\\azure-powershell\\src\\ResourceManager\\Aks\\Aks.sln":[
+        "Commands.Aks",
+        "Commands.Resources.Rest",
+        "Commands.Resources"
+    ],
+    "C:\\azure-powershell\\src\\ResourceManager\\AnalysisServices\\AnalysisServices.sln":[
+        "Commands.AnalysisServices",
+        "Commands.AnalysisServices.Dataplane"
+    ],
+    ...
+}
+#>
+function Create-SolutionToProjectMappings
+{
+    $Mappings = [ordered]@{}
+    foreach ($ServiceFolder in $Script:ServiceFolders)
+    {
+        $SolutionFiles = Get-ChildItem -Path $ServiceFolder.FullName -Filter "*.sln" | where { $_.FullName -notlike "*Stack*" -and $_.FullName -notlike "*Netcore*" }
+        foreach ($SolutionFile in $SolutionFiles)
+        {
+            $Mappings = Add-ProjectDependencies -Mappings $Mappings -SolutionPath $SolutionFile.FullName
+        }
+    }
+
+    $Mappings = Add-ProjectDependencies -Mappings $Mappings -SolutionPath (Join-Path -Path $Script:ServiceManagementPath -ChildPath "ServiceManagement.sln")
+    $Mappings = Add-ProjectDependencies -Mappings $Mappings -SolutionPath (Join-Path -Path $Script:StoragePath -ChildPath "Storage.sln")
+    return $Mappings
+}
+
+<#
+Parses a solution file to find the projects it is composed of (excluding test and common projects).
+#>
+function Add-ProjectDependencies
+{
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [hashtable]$Mappings,
+
+        [Parameter(Mandatory = $true)]
+        [string]$SolutionPath
+    )
+
+    $ProjectDependencies = @()
+    $Content = Get-Content -Path $SolutionPath
+    $Content | Select-String -Pattern "`"[a-zA-Z.]*`"" | % { $_.Matches[0].Value.Trim('"') } | where { $_ -notlike "Test*" -and $_ -notlike "*.Test*" -and $_ -notlike "*Common*" } | % { $ProjectDependencies += $_ }
```

Comment:
So, don't we want to specifically match ```<HintPath>path</HintPath>```, I worry that this match is a bit overly broad.

Reply:
This function matches projects (excluding test and common projects) built in a provided solution.

For example it will match `Commands.Compute` from the line 

```
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Compute", "Commands.Compute\Commands.Compute.csproj", "{52643BD5-6378-49BD-9F6E-DAC9DD8A867B}"
```